### ISSUE TITLE
Customized event handler

### ISSUE THREAD

**Issue by dukesun99** [neo4j-contrib/neovis.js#151]
I'm thinking of adding a customized event handler which can be used with the following syntax:
```js
"NodeLabel": {
			// other properties
                        "click": (values) => {
					document.getElementById("details").innerHTML = `</strong> `+ values.node.title;
                        },
                        "doubleclick": (values) => {
					document.getElementById("details").innerHTML = `</strong> `+ values.node.title;
                        },
}
```
And we can also add right-click displays that are not natively supported in the vis.js network but I think it would be good to have.

**Comment by thebestnom** [comment]
I did implemented this once like that, but full event was chosen, and Im not sure about adding 2 way for doing the (almost) same thing, need more opinion on that, that I wish I had

What do you think?

**Comment by dukesun99** [comment]
> I did implemented this once like that, but full event was chosen, and Im not sure about adding 2 way for doing the (almost) same thing, need more opinion on that, that I wish I had
> 
> 
> 
> What do you think?

thebestnom what do you mean by full event was chosen?

**Comment by thebestnom** [comment]
I mean that listening to the full div was added

See https://github.com/neo4j-contrib/neovis.js/issues/16

But it looks like I was misremembering, I saw that the PR for event was closed, so this us a great feature without the drawbacksI thought at the start

**Comment by dukesun99** [comment]
@thebestnom ok I will add events in the way I mentioned. If you have time maybe could you help to review my changes for the pr?

**Comment by thebestnom** [comment]
No problem!

**Comment by dukesun99** [comment]
I am also thinking of adding right click context menu as events, that could be a useful feature.

**Comment by thebestnom** [comment]
I think you should add this as a more generic any event from vis if you can think of a way to do so

**Comment by dukesun99** [comment]
> I think you should add this as a more generic any event from vis if you can think of a way to do so

@thebestnom I feel it would be hard to have a very clean way of doing this. Because some events do not return node id or edge id, in this case we cannot decide which handler to call. We can only manually register those events which are associated with a specific node or edge.

**Comment by dukesun99** [comment]
@thebestnom I am thinking of manually define all events we can support and there are not too many of them.

**Comment by thebestnom** [comment]
It does sounds like a pain... Ok, Ill look at the PR once you open it

**Comment by davidsbatista** [comment]
is there any advances on this feature? I would like to catch when a node or an edge is clicked, is this possible in the current master? or is there a branch where this is possible?

**Comment by thebestnom** [comment]
Yes, in 2.0.0-alpha.0 you can listen on click and then test to see what was clicked, this specific way is not implemented but it's in development and I think Ill continue from where he left

**Comment by davidsbatista** [comment]
thanks for the reply, but I just abandoned this `neovi.js`, I think currently this project is in a state where only a contributor/developer can actually use it properly; it lacks proper documentation and examples on how to use it, at least for me, and I found myself always going around in circles, building the lib using npm to see changes I've done, and failing to understand why things don't work;

I just started using visjs network from scratch together with Neo4J HTTP endpoint, and I've managed to get the same result as with neovis.js; but thanks for you all your help and messages; I will look into neovis.js again once it's in a more mature state;

**Comment by Wolfgang-TN** [comment]
Hallo all, please, **continue to use neovis**! 
I use it already from the beginning to show my visual ontology of mechanical calculating machines I collect (13000 nodes/27000 edges).
For all the click-listening etc., it is better you use **only the completed event**, where  you call a kind of handler function with the parameter 'viz._network'. There you can do all the special handlers you want. I have implemented even an onhold which results in showing only the directly connected nodes.
In the complete function you could call also a format function for each node/edge to change the visibility aspect. Eventually, you must tediously decompose the title property to obtain the single properties.:
`....
			let nodo=node.title.replace(/\:\/\//g,'%%').replace(/1\:8/g,'yy').replace(/\: /g,'-').replace(/\ :/g,'').replace(/:true/g,";true").replace(/Bild\:/,'Bild;').replace(/<strong>/g,'"').replace(/<\/strong>/g,'"').replace(/<br>/g,'",').replace(/\:/g,'":').replace(/yy/g,'1\:8').replace(/%%/g,"://").replace(/;true/g,":true").replace(/Bild;/,'Bild:');
			nodo="{"+nodo.substring(0,nodo.length-1)+"}";
			node=JSON.parse(nodo);
			node.id=item.id;
....
`
I'm new to github, but try to publish my code, soon. for the moment enjoy the attached screen scan.
![Neovis_calculators_WJI](https://user-images.githubusercontent.com/77717584/108484765-764aa480-729c-11eb-80a6-64dc0e57deaa.jpg)

Using Neo4j-Desktop 1.4.1/Neo4j 4.2.3, Neovis.v1.5.0, unpkg.com/neo4j-driver, and >4000 Javascript lines

**Comment by thebestnom** [comment]
@Wolfgang-TN thank you for being supportive 😄 but he is correct, for the current state of 2.0.0 alpha the docs are bad if you can't use typescript, Im working on it and its in the roadmap
My point with 2.0.0 is to make neovis way more versatile and allow to use it more tham to just see a single cypher but make it still as easy to use (I added `.network` as a getter to have the ability to do what you want with the network)

**Comment by Wolfgang-TN** [comment]
Hi thebestnorm, thanks for commenting my observations.
for a long time, I tried to mess directly the neovis code, but recently I prefer to use a published version (1.5.0 for now) adding my own requests externally. Of course, I had to use all the possible properties and methods of **visjs** from Javascript. As an example, here my **composed cypher code** which essentially executes my onhold: i.e. a call to hold({nodes:[9710]}):

`match (n) where id(n)=9710 with n optional match (n)<-[t]-(y) with n,t,y,id(n) as nodeId limit 1 optional match (n)-[r]-(m) with n,t,y,nodeId,r,m match (n) with n,t,y,nodeId,n.labels as myLabels,n.label as label,  count(m) as connected  call apoc.create.vNode([myLabels],{group:'_total',shape:'database',fromLabel:labels(n),fromName:label,from:nodeId,label:connected+'#',title:'hold to open all '+connected+' '+label+' nodes',count:connected}) yield node  call apoc.create.vRelationship(n,'HAS',{dashes:true,length:0.1}, node) YIELD rel  match (x) where id(x)=nodeId optional match (x)-[r]-(m)  with n,t,y,m,r, x, node, rel, connected  return n,t,y,m,r, x, CASE when connected>10 then node END,CASE when connected>10 then rel END  order by rand() limit 10`

the _9710_ is the id(n) of the node in question. Works with either node.id, but shows the ugly dot-nodes in Neo4j-Browser; if formated as mentioned before, the nodes may have all desired  visjs shapes. Think of using a similar cypher query for the onhold event

**Comment by thebestnom** [comment]
@Wolfgang-TN please open a new issue for continuing discussing this issue, as for 2.0.0, as I said here above, 2.0.0-alpha.0 is a published version on npm (there is an auto cdn for that that Ill like to test) and you can try using that

**Comment by Wolfgang-TN** [comment]
Hi thebestnom, to no avail,
your version 2.0.0 doesn't work:
1. Uncaught SyntaxError: Unexpected token '.'...neovis.js:98 
![image](https://user-images.githubusercontent.com/77717584/108543975-1aa50900-72e6-11eb-9194-76bb1eb3230e.png)
in Chrome  79.0.3945.117 (Build ufficiale) (a 64 bit), probably optional chaining (?.) not yet possible
2. Firefox still lacks private fields #network etc. (Chrome does)
hence I'll stay with the old well-working version. Let know when acceptable

**Comment by thebestnom** [comment]
published `2.0.0-alpha.2` that should fix both of that problems, default babel-preset behavior is wierd 😅

**Comment by thebestnom** [comment]
@davidsbatista sorry for the ping, I tried to make a new documention and I would like to get your input 😄
https://github.com/thebestnom/neovis.js/tree/feature/doc
